### Riddler Express

You and your two best friends are in a three-person league, drafting just three positions each for your teams: quarterback, running back and wide receiver. Yes, this is a simplified version of fantasy football.

The following table shows the top three athletes in each position, as well as the number of fantasy points they are expected to earn over the course of the season. You and your friends must each select exactly one player from each position.

The draft is a “snake draft.” If person A drafts first, B drafts second and C drafts third, then the order of the picks is as follows: A-B-C-C-B-A-A-B-C.

Your friends — being the kind people that they are — agree that you can choose your pick number. Which draft position should you choose to maximize your expected fantasy score?

### Assumptions:

- I will start by assuming everyone will take the top person available on the board. 

- Assume I am p1

In [1]:
from operator import itemgetter

# data, inspired by fluent python ch 7
data = [
    ('qb', 400),
    ('qb', 350),
    ('qb', 300),
    ('rb', 300),
    ('rb', 225),
    ('rb', 200),
    ('wr', 250),
    ('wr', 225),
    ('wr', 175),
]

# use itemgetter to sort list of tuples based on points
for player in sorted(data, key = itemgetter(1), reverse = True):
    print(player)


('qb', 400)
('qb', 350)
('qb', 300)
('rb', 300)
('wr', 250)
('rb', 225)
('wr', 225)
('rb', 200)
('wr', 175)


### If I Pick First ..

This actually doesn't work...ignores strategy. 

need to add logic in for "don't take a position if it is the last unless you have no other choice"

In [16]:
# data, inspired by fluent python ch 7
data = [
    ('qb', 400),
    ('qb', 350),
    ('qb', 300),
    ('rb', 300),
    ('rb', 225),
    ('rb', 200),
    ('wr', 250),
    ('wr', 225),
    ('wr', 175),
]

# dict of picks
p1, p2, p3 = {'picker': 'p1'}, {'picker': 'p2'}, {'picker': 'p3'}

# build order:
picks = [p1,p2,p3]
order = picks + picks[::-1] + picks

for pick in order:
    for player in sorted(data, key = itemgetter(1), reverse = True):
        
        # if we have this position already, move to the next
        if player[0] in pick.keys():
            next
        
        # otherwise, add to dict of picks, remove from data 
        pick[player[0]] = player[1]
        data.remove(player)
        
        break

# Final Scores:
for picker in [p1,p2,p3]:
    person = picker['picker']
    picker.pop('picker', None)
    print(f"{person} had a total score of {sum(picker.values())}")
    print(f"Picks were: {picker}")

p1 had a total score of 850
Picks were: {'qb': 400, 'rb': 225, 'wr': 225}
p2 had a total score of 800
Picks were: {'qb': 350, 'wr': 250, 'rb': 200}
p3 had a total score of 775
Picks were: {'qb': 300, 'rb': 300, 'wr': 175}


### Picking 3rd Makes Most Sense:  

Able to hold off on taking a QB until last pick. 

Strategy:

- let say we have `p2-p3-p1`
- Round 1: 
    - p2 - qb, 400
    - p3 - qb, 350
    - p1 - rb, 300 (since p2 and p3 can't take the 300 qb)
- round 2:
    - p1 - wr, 250
    - p3 - wr 225
    - p2 - rb, 225
- Round 3:
    - p2 - wr - 175
    - p3 - rb, 200
    - p1 - qb, 300
    
3rd pick -> P1 final score: 850

1st pick -> P2 final score: 800

2nd pick -> P3 final score: 775